File: 2018/01-frequency-changes/python/star2.py
#!/usr/bin/env python3

def lines_loop():
    while True:
        for line in lines:
            yield line


with open('input') as f:
    lines = f.readlines()

frequency_changes = map(int, lines_loop())

frequency = 0
seen_frequencies = set()
for change in frequency_changes:
    if frequency in seen_frequencies:
        break
    seen_frequencies.add(frequency)
    frequency += change

print(frequency)

File: 2018/01-frequency-changes/rust/star2.rs
use std::collections::HashSet;

fn main() {
    let input = include_str!("../input");

    let mut frequency = 0;
    let mut seen_frequencies = HashSet::new();
    'outer: loop {
        for line in input.lines() {
            let change = line.parse::<i32>().unwrap();
            frequency += change;
            if seen_frequencies.contains(&frequency) {
                println!("{}", frequency);
                break 'outer;
            }
            seen_frequencies.insert(frequency);
        }
    }
}
